Lib/test/test_shelve.py

```diff
@@ -165,6 +166,277 @@ def test_default_protocol(self):
+                        self.assertEqual(value, "str")
+
+                        key, value = s.previous()
+                        self.assertEqual("bytes_data", key)
+                        self.assertEqual(value, "bytes")
+
+                        key, value = s.previous()
+                        self.assertEqual("bytearray_data", key)
+                        self.assertEqual(value, "bytearray")
+
+                        key, value = s.previous()
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(value, "array")
+
+                        key, value = s.next()
+                        self.assertEqual("bytearray_data", key)
+                        self.assertEqual(value, "bytearray")
+
+                        key, value = s.next()
+                        self.assertEqual("bytes_data", key)
+                        self.assertEqual(value, "bytes")
+
+                        key, value = s.first()
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(value, "array")
+
+                        self.assertEqual(s["foo"], "str")
+                        self.assertEqual(s["bytes_data"], "bytes")
+                        self.assertEqual(s["bytearray_data"], "bytearray")
+                        self.assertEqual(s["array_data"], "array")
+
+    def test_custom_incomplete_serializer_and_deserializer_bsd_db_shelf(self):
+        berkeleydb = import_helper.import_module("berkeleydb")
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        def serializer(obj, protocol=None):
+            return type(obj).__name__.encode("utf-8")
+
+        def deserializer(data):
+            pass
+
+        with shelve.BsdDbShelf(berkeleydb.btopen(self.fn),
+                               serializer=serializer,
+                               deserializer=deserializer) as s:
+            s["foo"] = "bar"
+            self.assertIsNone(s["foo"])
+            self.assertNotEqual(s["foo"], "bar")
+
+        def serializer(obj, protocol=None):
+            pass
+
+        def deserializer(data):
+            return data.decode("utf-8")
+
+        with shelve.BsdDbShelf(berkeleydb.btopen(self.fn),
+                               serializer=serializer,
+                               deserializer=deserializer) as s:
+            s["foo"] = "bar"
+            self.assertEqual(s["foo"], "")
+            self.assertNotEqual(s["foo"], "bar")
+
+
+    def test_missing_custom_deserializer(self):
+        def serializer(obj, protocol=None):
+            pass
+
+        with self.assertRaises(shelve.ShelveError):
+            shelve.Shelf({},
+                         protocol=2, writeback=False, serializer=serializer)
+
+        with self.assertRaises(shelve.ShelveError):
+            shelve.BsdDbShelf({},
+                              protocol=2,
+                              writeback=False, serializer=serializer)
```

Comment:
```suggestion
        kwargs = dict(protocol=2, writeback=False, serializer=serializer)
        self.assertRaises(shelve.ShelveError, shelve.Shelf, {}, **kwargs)
        self.assertRaises(shelve.ShelveError, shelve.BsdDbShelf, {}, **kwargs)
```

We can reduce a bit the "height" of the test that way.